merge.go

```diff
@@ -217,3 +249,31 @@ func shouldKeyMergeArray(obj []interface{}) (bool, string) {
 	}
 	return false, ""
 }
+
+func canKeyMergeArray(disp string, array []interface{}, node string, key string) error {
+	// ensure that all elements of `array` are maps,
+	// and that they contain the key `key`
+
+	for i, o := range array {
+		if reflect.TypeOf(o).Kind() != reflect.Map {
+			return fmt.Errorf("%s.%d: %s object is a %s, not a map - cannot merge using keys", node, i, disp, reflect.TypeOf(o).Kind().String())
+		}
+
+		obj := o.(map[interface{}]interface{})
+		if _, ok := obj[key]; !ok {
+			return fmt.Errorf("%s.%d: %s object does not contain the key '%s' - cannot merge", node, i, disp, key)
+		}
+	}
+	return nil
+}
+
+func shouldReplaceArray(obj []interface{}) bool {
```

Comment:
We should have unit tests for this function as well, to verify the regexes get matched right.. should be pretty copy/paste-able from the other shouldXXXX tests